ios/SingularBridge.m

```diff
@@ -0,0 +1,181 @@
+    }
+
+    [Singular startSession:apikey withKey:secret];
+}
+
+RCT_EXPORT_METHOD(initWithSingularLinks:(NSString*)apikey
+                  secret:(NSString*)secret
+                  customUserId:(NSString*)customUserId){
+    if(customUserId){
+        [Singular setCustomUserId:customUserId];
+    }
+
+    isSingularLinksEnabled = YES;
+    eventEmitter = self;
+
+    [Singular startSession:apikey
+                   withKey:secret
+          andLaunchOptions:launchOptions
+   withSingularLinkHandler:^(SingularLinkParams * params){
+        [SingularBridge handleSingularLinks:params];
+    }];
+}
+
+RCT_EXPORT_METHOD(setCustomUserId:(NSString*)customUserId){
+    [Singular setCustomUserId:customUserId];
+}
+
+RCT_EXPORT_METHOD(unsetCustomUserId){
+    [Singular unsetCustomUserId];
+}
+
+RCT_EXPORT_METHOD(event:(NSString*)eventName){
+    [Singular event:eventName];
+}
+
+RCT_EXPORT_METHOD(eventWithArgs:(NSString*)eventName args:(NSString*)args){
+    if(!args){
+        [Singular event:eventName];
+    }
+
+    [Singular event:eventName withArgs:[SingularBridge jsonToDictionary:args]];
+}
+
+RCT_EXPORT_METHOD(revenue:(NSString*)currency amount:(double)amount){
+    [Singular revenue:currency amount:amount];
+}
+
+RCT_EXPORT_METHOD(revenueWithArgs:(NSString*)currency amount:(double)amount args:(NSString*)args){
+    if(!args){
+        [Singular revenue:currency amount:amount];
+    }
+
+    [Singular revenue:currency amount:amount withAttributes:[SingularBridge jsonToDictionary:args]];
+}
+
+RCT_EXPORT_METHOD(customRevenue:(NSString*)eventName currency:(NSString*)currency amount:(double)amount){
+    [Singular customRevenue:eventName currency:currency amount:amount];
+}
+
+RCT_EXPORT_METHOD(customRevenueWithArgs:(NSString*)eventName currency:(NSString*)currency amount:(double)amount args:(NSString*)args){
+    if(!args){
+        [Singular customRevenue:eventName currency:currency amount:amount];
+    }
+
+    [Singular customRevenue:eventName currency:currency amount:amount withAttributes:[SingularBridge jsonToDictionary:args]];
+}
+
+RCT_EXPORT_METHOD(setUninstallToken:(NSString*)token){
+    [Singular registerDeviceTokenForUninstall:[token dataUsingEncoding:NSUTF8StringEncoding]];
+}
+
+RCT_EXPORT_METHOD(trackingOptIn){
+    [Singular trackingOptIn];
+}
+
+RCT_EXPORT_METHOD(trackingUnder13){
+    [Singular trackingUnder13];
+}
+
+RCT_EXPORT_METHOD(stopAllTracking){
+    [Singular stopAllTracking];
+}
+
+RCT_EXPORT_METHOD(resumeAllTracking){
+    [Singular resumeAllTracking];
+}
+
+RCT_REMAP_METHOD(isAllTrackingStopped, resolver: (RCTPromiseResolveBlock)resolve
+                 rejecter:(RCTPromiseRejectBlock)reject){
+    resolve([Singular isAllTrackingStopped] ? @YES : @NO);
+}
+
+RCT_EXPORT_METHOD(setReactSDKVersion:(NSString*)wrapper version:(NSString*)version){
+    [Singular setWrapperName:wrapper andVersion:version];
+}
+
+#pragma mark - Private methods
+
++(NSDictionary*)jsonToDictionary:(NSString*)json{
+    NSError *jsonError;
```

Comment:
afaik, best practice is to initialize any `NSError*` with `nil` explicitly and not rely on the compiler doing it...